### DISCUSSION TITLE
What does `KMathContext` do?

### DISCUSSION THREAD

**Discussion by lounres** [mipt-npm/kmath#487] [General]
There is annotation class `KMathContext` in `kmath-core/src/commonMain/kotlin/space/kscience/kmath/operations/Algebra.kt` that is not used anywhere. (All appearances of word "KMathContext" are only declarations of the class in `.kt` sources, `.api` dump file and diagram in docs.) Is there really any functionality behind it?

**Comment by altavir** [comment]
KMathContext is a marker for https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-dsl-marker/ functionality. Used to prohibit implicit scope propagation inside inner scope. Basically if for some reason we have two different scopes of the same type on inside another we do not want to occasionally call outer one. The `DslMarker` is not quite stright-forward, but I hope it will be clearer in future.

**Reply by lounres** [reply]
Well, I already know what DSL marker annotations are. (I have fully read Kotlin in Action, where they are well described.) The question is not about `DslMarker` but `KMathContext` (and functionality not of `DslMarker` but of `KMathContext`).

To use mechanism of DSL markers you must apply the markers at least once. But there is no usage of `KMathContext`. It's just declared, marked as DSL marker and not used. There is no actual appearance of the annotation in Kotline sources except declaration of the annotation class.

So I ask what there is unused annotation class for.

**Reply by altavir** [reply]
Obviously because somebody forgot to use it or used at some moment and then stopped using it. It happens with research libraries :)

**Comment by altavir** [comment]
An additional benefit is that functions marked with `DslMarker` are highlighted in Idea
